Allow `truncate` for SQLite3 adapter and add `rails db:seed:replant`

bogdanvlviv opened the pull request:
- Add `ActiveRecord::Base.connection.truncate` for SQLite3 adapter.
    SQLite doesn't support `TRUNCATE TABLE`, but SQLite3 adapter can support
    `ActiveRecord::Base.connection.truncate` by using `DELETE FROM`.
    `DELETE` without `WHERE` uses "The Truncate Optimization",
    see https://www.sqlite.org/lang_delete.html.

- Add `rails db:seed:replant` that truncates tables of each database for current environment and loads the seeds.
  Closes #34765

bogdanvlviv committed:
Add `ActiveRecord::Base.connection.truncate` for SQLite3 adapter.

SQLite doesn't support `TRUNCATE TABLE`, but SQLite3 adapter can support
`ActiveRecord::Base.connection.truncate` by using `DELETE FROM`.

`DELETE` without `WHERE` uses "The Truncate Optimization",
see https://www.sqlite.org/lang_delete.html.

bogdanvlviv committed:
Add `rails db:seed:replant` that truncates database tables and loads the seeds

Closes #34765

dhh commented:
Awesome, @bogdanvlviv 🙏. I'll let someone else review the implementation, but this is great.

simi commented:
What about to use also `CASCADE` to not crash on foreign key constraints?

bogdanvlviv commented:
> What about to use also CASCADE to not crash on foreign key constraints?

`ActiveRecord::Base.connection.truncate` doesn't allow to set "CASCADE", so I've tried to use `ActiveRecord::Base.connection.disable_referential_integrity`. It solved this issue for MySQL2, but for PostgreSQL it is still issue :thinking:.



```diff
+++ b/activerecord/lib/active_record/railties/databases.rake
@@ -68,8 +68,10 @@ db_namespace = namespace :db do
       ActiveRecord::Base.internal_metadata_table_name
     ]

-    table_names.without(*internal_table_names).each do |table_name|
-      ActiveRecord::Base.connection.truncate(table_name)
+    ActiveRecord::Base.connection.disable_referential_integrity do
+      table_names.without(*internal_table_names).each do |table_name|
+        ActiveRecord::Base.connection.truncate(table_name)
+      end
     end
   end
```

bogdanvlviv committed:
Add `ActiveRecord::Tasks::DatabaseTasks.truncate_tables`

Extract code from the rake task. It will be easier to test this code and reuse.

bogdanvlviv committed:
Apply `ActiveRecord::Base.connection.disable_referential_integrity` when `truncate_tables`

bogdanvlviv commented:
On PostgreSQL, [this issue](https://github.com/rails/rails/pull/34779#issuecomment-450208705) ([test failure](https://travis-ci.org/rails/rails/jobs/472966915#L1294)) could be solved by applying `CASCADE`. We could change the `truncate` method to allow pass additional options  like `ActiveRecord::Base.connection.truncate(table_name, nil, cascade: true)`, see diff:
<details>
<summary>Diff</summary>
<br>

```diff
diff --git a/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb b/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb
index 2299fc0214..deac8e9e5b 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb
@@ -135,7 +135,7 @@ def exec_delete(sql, name = nil, binds = [])
       end

       # Executes the truncate statement.
-      def truncate(table_name, name = nil)
+      def truncate(table_name, name = nil, options = {})
         raise NotImplementedError
       end

diff --git a/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb b/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb
index dbc6614b93..ad13432aaa 100644
--- a/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb
+++ b/activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb
@@ -271,7 +271,7 @@ def collation
         show_variable "collation_database"
       end

-      def truncate(table_name, name = nil)
+      def truncate(table_name, name = nil, options = {})
         execute "TRUNCATE TABLE #{quote_table_name(table_name)}", name
       end

diff --git a/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb b/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
index 381d5ab29b..7ed0c45c17 100644
--- a/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
+++ b/activerecord/lib/active_record/connection_adapters/postgresql_adapter.rb
@@ -258,8 +258,11 @@ def clear_cache!
         end
       end

-      def truncate(table_name, name = nil)
-        exec_query "TRUNCATE TABLE #{quote_table_name(table_name)}", name, []
+      def truncate(table_name, name = nil, options = {})
+        cascade = if options[:cascade]
+          " CASCADE"
+        end
+        exec_query "TRUNCATE TABLE #{quote_table_name(table_name)}#{cascade}", name, []
       end

       # Is this connection alive and ready for queries?
diff --git a/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb b/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb
index 94754e5b86..144d1548b9 100644
--- a/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb
+++ b/activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb
@@ -161,7 +161,7 @@ def clear_cache!
         @statements.clear
       end

-      def truncate(table_name, name = nil)
+      def truncate(table_name, name = nil, options = {})
         execute "DELETE FROM #{quote_table_name(table_name)}", name
       end

diff --git a/activerecord/lib/active_record/tasks/database_tasks.rb b/activerecord/lib/active_record/tasks/database_tasks.rb
index 81a2b84e39..f852ebf328 100644
--- a/activerecord/lib/active_record/tasks/database_tasks.rb
+++ b/activerecord/lib/active_record/tasks/database_tasks.rb
@@ -191,7 +191,7 @@ def truncate_tables

         ActiveRecord::Base.connection.disable_referential_integrity do
           table_names.without(*internal_table_names).each do |table_name|
-            ActiveRecord::Base.connection.truncate(table_name)
+            ActiveRecord::Base.connection.truncate(table_name, nil, cascade: true)
           end
         end
       end
```
</details>

But I'm not completly sure that it is right solution since it changes public API(`truncate` method).
Do you have any suggestion about solving this?

bogdanvlviv committed:
`rails db:seed:replant` should work with multi DB

Since version 6.0 Rails will support mult DB, so user will be
able to write in `db/seeds.rb` something like:

```ruby
ActiveRecord::Base.connected_to(database: :primary) do
  User.create!(name: "Bogdan")
end

ActiveRecord::Base.connected_to(database: :animals) do
  Dog.create!(name: "Bosyi")
end
```

`rails db:seed:replant` should truncate tables in both databases,
and replant seeds.

</hr>

Since not every DB backend has `TRUNCATE`, or if it has then
it can work in a different way than on any another DB backend.

For instance, in order to truncate a table and not crash on foreign key constraints
on MySQL we should execute something like:

```sql
SET FOREIGN_KEY_CHECKS = 0;
TRUNCATE TABLE `users`;
```

on PostgreSQL:

```sql
TRUNCATE TABLE "users", ... ;
```

on SQLite:

```sql
PRAGMA defer_foreign_keys = ON;
PRAGMA foreign_keys = OFF;
DELETE FROM "users";
```

Because of that, we should also provide strategy by which other DB
adapters, like `oracle-enhanced`, or `activerecord-sqlserver-adapter`,
will we able to make `rails db:seed:replant` work.

Since we can't just rely on `ActiveRecord::Base.connection.truncate`
because of the inconsistency of DB backends, I want to recommend the next:
Implement "ActiveRecord::Tasks::#{adapter_name}DatabaseTasks#truncate_tables(*table_names)"
to an adapter that would like to support this rake task.
This method just receives an array of table names. Your implementation
should ensure that it truncates all of those tables. Also, be sure
that it does not crash on foreign key constraints!

bogdanvlviv commented:
`rails db:seed:replant` should work with multi DB as well

Since version 6.0 Rails will support mult DB, so user will be
able to write in `db/seeds.rb` something like:

```ruby
ActiveRecord::Base.connected_to(database: :primary) do
  User.create!(name: "Bogdan")
end

ActiveRecord::Base.connected_to(database: :animals) do
  Dog.create!(name: "Bosyi")
end
```

`rails db:seed:replant` should truncate tables in both databases,
and replant seeds.

</hr>

Since not every DB backend has `TRUNCATE`, or if it has then
it can work in a different way than on any another DB backend.

For instance, in order to truncate a table and not crash on foreign key constraints
on MySQL we should execute something like:

```sql
SET FOREIGN_KEY_CHECKS = 0;
TRUNCATE TABLE `users`;
```

on PostgreSQL:

```sql
TRUNCATE TABLE "users" CASCADE;
```

on SQLite:

```sql
PRAGMA defer_foreign_keys = ON;
PRAGMA foreign_keys = OFF;
DELETE FROM "users";
```

Because of that, we should also provide strategy by which other DB
adapters, like `oracle-enhanced`, or `activerecord-sqlserver-adapter`,
will we able to make `rails db:seed:replant` work.

Since we can't just rely on `ActiveRecord::Base.connection.truncate`
because of the inconsistency of DB backends, I want to recommend the next:
Implement "ActiveRecord::Tasks::#{adapter_name}DatabaseTasks#truncate_tables(*table_names)"
to an adapter that would like to support this rake task.
This method just receives an array of table names. Your implementation
should ensure that it truncates all of those tables. Also, be sure
that it does not crash on foreign key constraints!

/cc @dhh

matthewd commented:
> Since we can't just rely on `ActiveRecord::Base.connection.truncate`

Yes, we can: to be compatible with a new release, adapters need to support whatever methods we decide on. (And our downstreams are super accomodating and responsive about that.)

I haven't looked at this, but we already have "empty out all the tables" behaviour for populating fixtures: unless I'm missing something, that should both prove that we have the technology, and suggest some scope for code sharing.

matthewd commented:
> I've only found this

Yeah, that looks right: it disables referential integrity, then empties out all the tables.

We could consider switching that to truncate, if its additional constraints are okay, but I think we can treat that as a separate question.

(As a side note, `CASCADE` is not a solution here: by definition it means something's going to get cleared out that we're not expecting, which would be unhelpful -- in contrast, listing all the known tables in a single TRUNCATE statement might be useful)

bogdanvlviv commented:
> (As a side note, CASCADE is not a solution here: by definition it means something's going to get cleared out that we're not expecting, which would be unhelpful -- in contrast, listing all the known tables in a single TRUNCATE statement might be useful)

:+1:  Thanks for pointing this out, I changed the implementation for PostgreSQL to use TRUNCATE statement with a listing of known tables.

kaspth commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -182,6 +182,24 @@ def drop_current(environment = env)
+        table_names = ActiveRecord::Base.connection.tables
+        internal_table_names = [
+          ActiveRecord::Base.schema_migrations_table_name,
+          ActiveRecord::Base.internal_metadata_table_name
```
Seems like this might have a need for something similar to what we're doing here: https://github.com/rails/rails/pull/33985/files#diff-18d566ee525de376b1a3741881545631R147

bogdanvlviv replied:
There is one more place that possible should be refactored
https://github.com/rails/rails/blob/84718df86097442f85999d6f2e6f6b8b59724c3f/activerecord/lib/active_record/schema_dumper.rb#L256. I'm not sure that we can use something like `ActiveRecord::Base.descendants.select { |model| !model._internal? }.map(&:table_name).compact` here (It ignores join tables, ...) since  It isn't the same like `ActiveRecord::Base.connection.tables`.

dhh commented:
@bogdanvlviv Is this ready for another review?

bogdanvlviv commented:
Yes. I just rebased with the `master` to resolve the conflicts.

dhh commented:
@kaspth and @matthewd, can you guys do a final review?

kaspth commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -182,6 +182,24 @@ def drop_current(environment = env)
+        ActiveRecord::Base.configurations.configs_for(env_name: environment).each do |db_config|
+          truncate_tables db_config.config
+        end
+        ActiveRecord::Base.establish_connection(environment.to_sym)
```
Isn't establish connection deprecated now that we have connect_to?

kaspth reviewed (commented):
Got one question, then I'll pass this on to @matthewd or whoever else wants to take this home.

bogdanvlviv committed:
Use `connected_to` instead of `establish_connection` on `rails db:seed:replant`

bogdanvlviv commented on activerecord/lib/active_record/tasks/database_tasks.rb:
```diff
@@ -182,6 +182,24 @@ def drop_current(environment = env)
+      end
+
+      def truncate_all(environment = env)
+        ActiveRecord::Base.configurations.configs_for(env_name: environment).each do |db_config|
```
I'm quite not sure whether we should truncate replica databases? (We should pass `include_replicas: true` to the `configs_for` if we want to do it).

bogdanvlviv commented:
I just rebased this branch one more time to resolve the conflicts. Let me if you have any concerns about these changes.

dhh commented:
@kaspth @matthewd Any other implementation notes?

dhh commented:
@bogdanvlviv You're happy with the state of things and ready to merge?

bogdanvlviv commented:
Unless there are any new implementation notes, it should be good to go.
One thing that concerns me is CI builds. Seems like there are some flaky tests (I just run tests on my VM and haven't got those test failures).
https://buildkite.com/rails/rails/builds/59209,
https://travis-ci.org/rails/rails/builds/501434510

dhh commented:
CI is terribly flaky at the moment 👎

Commit:
Merge branch 'master' into issue-34765

melodywei861016 commented:
I approve

melodywei861016 commented:
That's right
